// @progress/kendo-angular-grid/dist/es2015/editing/save-command.directive.js
import { Directive, HostListener, HostBinding, Inject, ElementRef, Renderer2 as Renderer } from '@angular/core';
import { Button } from '@progress/kendo-angular-buttons';
import { LocalizationService } from '@progress/kendo-angular-l10n';
import { EditService } from './edit.service';
import { CELL_CONTEXT } from '../rendering/common/cell-context';
/**
 * Represents the `save` command of the Grid. You can apply this directive to any `button`
 * element inside a [`CommandColumnComponent`]({% slug api_grid_commandcolumncomponent %}).
 * When an associated button with the directive is clicked, the
 * [`save`]({% slug api_grid_gridcomponent %}#toc-save) event
 * is triggered ([see example]({% slug editing_grid %})).
 *
 * > When the row is not in the edit mode, the button with `kendoGridSaveCommand` is automatically hidden.
 *
 * @example
 * ```html-no-run
 * <kendo-grid>
 *   <kendo-grid-command-column title="command">
 *     <ng-template kendoGridCellTemplate>
 *       <button kendoGridSaveCommand>Save changes</button>
 *     </ng-template>
 *   </kendo-grid-command-column>
 * </kendo-grid>
 * ```
 *
 * You can control the content of the button based on the state of the row.
 *
 * @example
 * ```html-no-run
 * <kendo-grid>
 *   <kendo-grid-command-column title="command">
 *     <ng-template kendoGridCellTemplate let-isNew="isNew">
 *       <button kendoGridSaveCommand>{{isNew ? 'Add' : 'Update'}}</button>
 *     </ng-template>
 *   </kendo-grid-command-column>
 * </kendo-grid>
 * ```
 */
export class SaveCommandDirective extends Button {
    constructor(editService, cellContext, element, renderer, localization) {
        super(element, renderer, null, localization);
        this.editService = editService;
        this.rowIndex = cellContext.rowIndex;
    }
    /**
     * @hidden
     */
    click() {
        this.editService.save(this.rowIndex);
    }
    /**
     * @hidden
     */
    get visible() {
        return !this.editService.isEdited(this.rowIndex) ? 'none' : '';
    }
    /**
     * @hidden
     */
    get commandClass() {
        return true;
    }
}
SaveCommandDirective.decorators = [
    { type: Directive, args: [{
                selector: '[kendoGridSaveCommand]'
            },] },
];
/** @nocollapse */
SaveCommandDirective.ctorParameters = () => [
    { type: EditService, },
    { type: undefined, decorators: [{ type: Inject, args: [CELL_CONTEXT,] },] },
    { type: ElementRef, },
    { type: Renderer, },
    { type: LocalizationService, },
];
SaveCommandDirective.propDecorators = {
    'click': [{ type: HostListener, args: ['click',] },],
    'visible': [{ type: HostBinding, args: ['style.display',] },],
    'commandClass': [{ type: HostBinding, args: ['class.k-grid-save-command',] },],
};
