// @progress/kendo-angular-inputs/dist/npm/common/draggable.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var dom_utils_1 = require("./dom-utils");
var kendo_draggable_1 = require("@telerik/kendo-draggable");
/**
 * @hidden
 */
var KendoDraggableDirective = /** @class */ (function () {
    function KendoDraggableDirective(ngEl) {
        var _this = this;
        this.kendoDrag = new core_1.EventEmitter();
        this.kendoPress = new core_1.EventEmitter();
        this.kendoRelease = new core_1.EventEmitter();
        if (!dom_utils_1.isDocumentAvailable()) {
            return;
        }
        this.draggable = new kendo_draggable_1.default({
            drag: function (e) {
                if (e.hasOwnProperty("originalEvent")) {
                    e.originalEvent.preventDefault();
                }
                _this.kendoDrag.next(e);
            },
            press: function (e) { return _this.kendoPress.next(e); },
            release: function (e) { return _this.kendoRelease.next(e); }
        });
        this.draggable.bindTo(ngEl.nativeElement);
    }
    KendoDraggableDirective.prototype.ngOnDestroy = function () {
        if (this.draggable) {
            this.draggable.destroy();
        }
    };
    KendoDraggableDirective.decorators = [
        { type: core_1.Directive, args: [{
                    selector: '[kendoDraggable]'
                },] },
    ];
    /** @nocollapse */
    KendoDraggableDirective.ctorParameters = function () { return [
        { type: core_1.ElementRef, },
    ]; };
    KendoDraggableDirective.propDecorators = {
        'kendoDrag': [{ type: core_1.Output },],
        'kendoPress': [{ type: core_1.Output },],
        'kendoRelease': [{ type: core_1.Output },],
    };
    return KendoDraggableDirective;
}());
exports.KendoDraggableDirective = KendoDraggableDirective;
