// @progress/kendo-angular-grid/dist/npm/excel/excel.component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var Observable_1 = require("rxjs/Observable");
var take_1 = require("rxjs/operators/take");
var kendo_file_saver_1 = require("@progress/kendo-file-saver");
var kendo_angular_excel_export_1 = require("@progress/kendo-angular-excel-export");
var excel_service_1 = require("./excel.service");
var excel_export_event_1 = require("./excel-export-event");
var kendo_angular_l10n_1 = require("@progress/kendo-angular-l10n");
var kendo_data_query_1 = require("@progress/kendo-data-query");
var column_common_1 = require("../columns/column-common");
/* tslint:disable object-literal-sort-keys */
var fetchComponentData = function (component) {
    return {
        data: component.view.map(function (item) { return item; }),
        group: component.group
    };
};
var toExcelColumn = function (column) {
    return {
        title: column.title,
        field: column.field,
        locked: Boolean(column.locked),
        width: column.width,
        level: column.level,
        hidden: !column.isVisible,
        groupHeaderTemplate: column.groupHeaderTemplate,
        groupFooterTemplate: column.groupFooterTemplate,
        footerTemplate: column.footerTemplate
    };
};
var toExcelColumns = function (columns) {
    var result = [];
    column_common_1.sortColumns(columns)
        .forEach(function (column) {
        if (column.childColumns) {
            result.push.apply(result, toExcelColumns(column.childColumns.toArray()));
        }
        else {
            var excelColumn = toExcelColumn(column);
            if (column.children) {
                excelColumn.children = [excelColumn].concat(toExcelColumns(column.children.toArray().slice(1)));
            }
            result.push(excelColumn);
        }
    });
    return result;
};
var componentColumns = function (component) {
    var columns = toExcelColumns(component.columns.toArray());
    return kendo_data_query_1.orderBy(columns, [{ field: 'locked', dir: 'desc' }]);
};
/**
 * Configures the settings for the export of Grid in Excel ([see example]({% slug excelexport_grid %})).
 */
var ExcelComponent = /** @class */ (function () {
    function ExcelComponent(excelService, localization) {
        this.localization = localization;
        /**
         * Specifies the file name of the exported Excel file.
         * @default "Export.xlsx"
         */
        this.fileName = 'Export.xlsx';
        /**
         * @hidden
         */
        this.columns = new core_1.QueryList();
        this.saveSubscription = excelService.saveToExcel.subscribe(this.save.bind(this));
    }
    ExcelComponent.prototype.ngOnDestroy = function () {
        this.saveSubscription.unsubscribe();
        if (this.dataSubscription) {
            this.dataSubscription.unsubscribe();
        }
    };
    ExcelComponent.prototype.save = function (component) {
        var _this = this;
        var data = (this.fetchData || fetchComponentData)(component);
        var exportData = function (result) {
            delete _this.dataSubscription;
            _this.exportData(component, result);
        };
        if (data instanceof Promise) {
            data.then(exportData);
        }
        else if (data instanceof Observable_1.Observable) {
            this.dataSubscription = data.pipe(take_1.take(1)).subscribe(exportData);
        }
        else {
            exportData(data);
        }
    };
    ExcelComponent.prototype.exportData = function (component, result) {
        var options = kendo_angular_excel_export_1.workbookOptions({
            columns: this.columns.length ? this.columns : componentColumns(component),
            data: result.data,
            group: result.group,
            filterable: this.filterable,
            creator: this.creator,
            date: this.date,
            paddingCellOptions: this.paddingCellOptions,
            headerPaddingCellOptions: this.headerPaddingCellOptions,
            rtl: this.localization.rtl
        });
        var args = new excel_export_event_1.ExcelExportEvent(options);
        component.excelExport.emit(args);
        if (!args.isDefaultPrevented()) {
            this.saveFile(options);
        }
    };
    ExcelComponent.prototype.saveFile = function (options) {
        var _this = this;
        kendo_angular_excel_export_1.toDataURL(options).then(function (dataURL) {
            kendo_file_saver_1.saveAs(dataURL, _this.fileName, {
                forceProxy: _this.forceProxy,
                proxyURL: _this.proxyURL
            });
        });
    };
    ExcelComponent.decorators = [
        { type: core_1.Component, args: [{
                    selector: 'kendo-grid-excel',
                    template: ""
                },] },
    ];
    /** @nocollapse */
    ExcelComponent.ctorParameters = function () { return [
        { type: excel_service_1.ExcelService, },
        { type: kendo_angular_l10n_1.LocalizationService, },
    ]; };
    ExcelComponent.propDecorators = {
        'fileName': [{ type: core_1.Input },],
        'filterable': [{ type: core_1.Input },],
        'creator': [{ type: core_1.Input },],
        'date': [{ type: core_1.Input },],
        'forceProxy': [{ type: core_1.Input },],
        'proxyURL': [{ type: core_1.Input },],
        'fetchData': [{ type: core_1.Input },],
        'paddingCellOptions': [{ type: core_1.Input },],
        'headerPaddingCellOptions': [{ type: core_1.Input },],
        'columns': [{ type: core_1.ContentChildren, args: [kendo_angular_excel_export_1.ColumnBase, { descendants: true },] },],
    };
    return ExcelComponent;
}());
exports.ExcelComponent = ExcelComponent;
